exercises/joseph_torsney/ex01_basics/src/main.cpp

```diff
@@ -0,0 +1,185 @@
+#include <iostream> 
+#include <fstream>
+#include <vector>
+#include <map>
+#include <algorithm>
+#include <string>
+
+/** Checks whether a character is punctuation
+ * 
+ * @param c 
+ * @return true if c is punctuation
+ * @return false otherwise
+ */
+bool myIsPunct(char c) {
+  return(
+    c == '.' ||
+    c == ',' ||
+    c == '?' ||
+    c == '\'' ||
+    c == '\"' ||
+    c == '!' ||
+    c == ')' ||
+    c == '('
+  );
+}
+
+/** Transforms a string to lower case.
+ * 
+ * @param str the string to transform
+ * @return std::string the string in lower case
+ */
+std::string toLowerStr(std::string str) {
+  std::transform(str.begin(), str.end(), str.begin(), 
+    [](unsigned char c){ return std::tolower(c); }
+  );
+
+  return str;
+}
+
+/** Reads the contents of a .txt file at filepath
+ * 
+ * @param filepath the path to the file
+ * @return std::string of the file contents
+ */
+std::string readFile(const std::string& filepath)
+{
+  std::ifstream ifs(filepath);
+  std::string text(
+    (std::istreambuf_iterator<char>(ifs)),
+    (std::istreambuf_iterator<char>())
+  );
+  
+  return(text);
+}
+
+/** Removes punctuation from a string
+ * 
+ * @param str the string to process
+ * @return std::string with no punctuation
+ */
+std::string removePunct(std::string str)
+{
+  for (int i = 0, len = (int) str.size(); i < len; i++) {
+    if (myIsPunct(str[i])) {
+      // if this char is punctuation, reduce i by 1, remove it, and
+      // reevaluate the string size.
+      str.erase(i--, 1);
+      len = (int) str.size();
+    }
+  }
+  return str;
+}
+
+/** Splits a string by spaces and hyphens into a string vector.
+ * 
+ * @param s the string to split
+ * @return std::vector<std::string> of 'words' contained in s.
+ */
+std::vector<std::string> split(const std::string& s)
+{
+  std::vector<std::string> ret;
+  typedef std::string::size_type string_size;
+  string_size i = 0;
+
+  // process characters of s
+  while (i != s.size()) {
+
+    // ignore whitespace
+    while(i != s.size() && (isspace(s[i]) || s[i] == '-')) {
+      i++;
+    }
+
+    // otherwise, we are at the start of a word.
+    string_size j = i;
```

Comment:
This loop performs as required functionally but in many cases it's possible to avoid relying on any complicated checks on indices in loops by using the c++ library a bit more.
In this case if you choose to read in the filestream word by word rather than doing it all at once in readFile you get a marker for free in the filestream which is equivalent to "j"
For example if you just open the filestream in readFile then there are a couple of approaches you can follow to simplify this function:
a) the >> operator respects the same logic as isspace and gives you the content of the file word by word. To manage the hyphens you then need to have an inner loop that uses the getline function. So something like:
```
while (filestr>>word){
  std::istringstream ss(wordWithoutHyphens);
  getline(filestr,wordWithoutHyphens,'-'){
    \\build up vector of words in here...
  }
}
```
b) if you first replace all hypens with a space then you don't need to bother with the getline inner loop